internal/lsp/object_stream.go

```diff
@@ -0,0 +1,92 @@
+package lsp
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/textproto"
+	"strconv"
+	"sync"
+
+	"github.com/formancehq/numscript/internal/jsonrpc2"
+)
+
+type LsObjectStream struct {
+	mu     sync.Mutex
+	reader *bufio.Reader
+
+	in  io.ReadCloser
+	out io.WriteCloser
+}
+
+var _ jsonrpc2.MessageStream = (*LsObjectStream)(nil)
+
+func NewLsObjectStream(in io.ReadCloser, out io.WriteCloser) LsObjectStream {
+	reader := bufio.NewReader(in)
+	return LsObjectStream{
+		reader: reader,
+		in:     in,
+		out:    out,
+	}
+}
+
+func (s *LsObjectStream) Close() error {
+	err := s.out.Close()
+	err2 := s.in.Close()
+
+	if err != nil {
+		return err
+	}
+
+	if err2 != nil {
+		return err2
+	}
+
+	return nil
+}
+
+func (s *LsObjectStream) WriteMessage(obj jsonrpc2.Message) error {
+	bytes, err := json.Marshal(obj)
```

Comment:
not a big deal but you can avoid some memory consumtion by streaming the marshaling.
```
json.NewEncoder(s.out).Encode(obj)
```